gpii/node_modules/preferencesServer/src/preferencesService.js

```diff
@@ -73,12 +124,273 @@ var fluid = fluid || require("infusion");
+        return promiseTogo;
+    };
+
+    /**
+     * Adds a new preferences safe to CouchDB using the preferences safe format.
+     *
+     * @param {Object} that - Instance of `gpii.preferencesServer.preferencesService`.
+     * @param {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param {Object} prefsSafeData - Preferences safe to add to system. The following fields
+     * are not required as part of this payload, and will be automatically created upon add:
+     * `id`, `timestampCreated`, and `timestampUpdated` properties.
+     * @return {Promise} Returns a promise resolving with the newly created
+     * preferences safe, or rejecting with an error payload containing an error `message`.
+     */
+    gpii.preferencesServer.preferencesService.addPrefsSafe = function (that, dataStore, prefsSafeData) {
+        var promiseTogo = fluid.promise();
+        if (prefsSafeData.type !== "prefsSafe") {
+            promiseTogo.reject({
+                isError: true,
+                errorCode: gpii.preferencesServer.errors.errorNotPrefsSafe.errorCode,
+                message: gpii.preferencesServer.errors.errorNotPrefsSafe.message
+            });
+        }
+        else {
+            // TODO This will have json schema validation when the scaffolding is all in.
+            dataStore.addPrefsSafe(prefsSafeData).then(
+                function (updatedReturn) {
+                    that.getPrefsSafe(updatedReturn.id).then(promiseTogo.resolve, promiseTogo.reject);
+                },
+                promiseTogo.reject);
+        };
+        return promiseTogo;
+    };
+
+    /**
+     * Updates a preferences safe in CouchDB.
+     *
+     * @param {Object} that - Instance of `gpii.preferencesServer.preferencesService`.
+     * @param {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param {Object} prefsSafeData - Preferences safe to add to system.
+     * @return {Promise} Returns a promise resolving with the newly created
+     * preferences safe, or rejecting with an error payload containing an error `message`.
+     */
+    gpii.preferencesServer.preferencesService.updatePrefsSafe = function (that, dataStore, prefsSafeData) {
+        var promiseTogo = fluid.promise();
+        if (!prefsSafeData.id || prefsSafeData.type !== "prefsSafe") {
+            promiseTogo.reject({
+                isError: true,
+                errorCode: gpii.preferencesServer.errors.errorNotPrefsSafeMissingId.errorCode,
+                message: gpii.preferencesServer.errors.errorNotPrefsSafeMissingId.message
+            });
+        }
+        else {
+            that.getPrefsSafe(prefsSafeData.id).then(function () {
+                // TODO This will have json schema validation when the scaffolding is all in.
+                dataStore.updatePrefsSafe(prefsSafeData.id, prefsSafeData).then(
+                    function (/* updatedReturn */) {
+                        that.getPrefsSafe(prefsSafeData.id).then(
```

Comment:
This is a too deeply nested promise chain and needs to be broken up

Reply:
This has been converted to a transforming promise chain.